extern/iso_c_fortran_bmi/src/bmi.f90

```diff
@@ -0,0 +1,569 @@
+  integer, parameter :: BMI_MAX_TYPE_NAME = 2048
+  integer, parameter :: BMI_MAX_UNITS_NAME = 2048
+  integer, parameter :: BMI_MAX_LOCATION_NAME = 2048
+  !Add a bound to max allowable file name length
+  integer, parameter :: BMI_MAX_FILE_NAME = 2048
+
+  integer, parameter :: BMI_FAILURE = 1
+  integer(kind=c_int), parameter :: BMI_SUCCESS = 0
+
+  type, abstract :: bmi
+    contains
+
+      ! Initialize, run, finalize (IRF)
+      procedure(bmif_initialize), deferred :: initialize
+      procedure(bmif_update), deferred :: update
+      procedure(bmif_update_until), deferred :: update_until
+      procedure(bmif_finalize), deferred :: finalize
+
+      ! Exchange items
+      procedure(bmif_get_component_name), deferred :: get_component_name
+      procedure(bmif_get_input_item_count), deferred :: get_input_item_count
+      procedure(bmif_get_output_item_count), deferred :: get_output_item_count
+      procedure(bmif_get_input_var_names), deferred :: get_input_var_names
+      procedure(bmif_get_output_var_names), deferred :: get_output_var_names
+
+!       ! Variable information
+      procedure(bmif_get_var_grid), deferred :: get_var_grid
+      procedure(bmif_get_var_type), deferred :: get_var_type
+      procedure(bmif_get_var_units), deferred :: get_var_units
+      procedure(bmif_get_var_itemsize), deferred :: get_var_itemsize
+      procedure(bmif_get_var_nbytes), deferred :: get_var_nbytes
+      procedure(bmif_get_var_location), deferred :: get_var_location
+
+!       ! Time information
+      procedure(bmif_get_current_time), deferred :: get_current_time
+      procedure(bmif_get_start_time), deferred :: get_start_time
+      procedure(bmif_get_end_time), deferred :: get_end_time
+       procedure(bmif_get_time_units), deferred :: get_time_units
+      procedure(bmif_get_time_step), deferred :: get_time_step
+! !
+!       ! Getters, by type
+      procedure(bmif_get_value_int), deferred :: get_value_int
+      procedure(bmif_get_value_float), deferred :: get_value_float
+      procedure(bmif_get_value_double), deferred :: get_value_double
+! !       procedure(bmif_get_value_ptr_int), deferred :: get_value_ptr_int
+! !       procedure(bmif_get_value_ptr_float), deferred :: get_value_ptr_float
+! !       procedure(bmif_get_value_ptr_double), deferred :: get_value_ptr_double
+! !       procedure(bmif_get_value_at_indices_int), deferred :: &
+! !            get_value_at_indices_int
+! !       procedure(bmif_get_value_at_indices_float), deferred :: &
+! !            get_value_at_indices_float
+! !       procedure(bmif_get_value_at_indices_double), deferred :: &
+! !            get_value_at_indices_double
+! !
+! !       ! Setters, by type
+      procedure(bmif_set_value_int), deferred :: set_value_int
+      procedure(bmif_set_value_float), deferred :: set_value_float
+      procedure(bmif_set_value_double), deferred :: set_value_double
+! !       procedure(bmif_set_value_at_indices_int), deferred :: &
```

Comment:
Are we certian that we don't need the 'set_value_at_indices()' functions

Reply:
We don't need them for the immediate need of using NoahMP fortran module.  We can extend this library with implementations for the unimplemented functions in the future.  Ideally, we should proxy each BMI function, and then any model that doesn't want to implement a particular function will still need to define it and simply always return `BMI_FAILURE`.